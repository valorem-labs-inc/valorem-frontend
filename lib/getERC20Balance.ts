import { Contract, providers, Signer } from "ethers";

import erc20ABI from "./abis/erc20";

export default async function getERC20Balance(
  address: string,
  account: string,
  provider: Signer
) {
  const contract = new Contract(address, erc20ABI, provider);

  return contract.balanceOf(address);
}
