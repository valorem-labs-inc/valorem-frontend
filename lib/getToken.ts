import tokens from "./tokens";
import { Token } from "./types";

const getToken = (address = ""): Token => {
  const searchForAddress = address.toLowerCase();
  return tokens.find((item) => item.address.toLowerCase() === searchForAddress);
};

export default getToken;
