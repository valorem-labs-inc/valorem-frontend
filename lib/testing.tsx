import * as React from "react";
import { jsonRpcProvider } from "wagmi/providers/jsonRpc";
import { render, RenderOptions } from "@testing-library/react";
import { Wallet, providers } from "ethers";
import {
  chain,
  configureChains,
  createClient,
  useAccount,
  useConnect,
  WagmiConfig,
} from "wagmi";
import { MockConnector } from "@wagmi/core/connectors/mock";

export const AutoConnect: React.FC = ({ children }) => {
  const { isConnected } = useAccount();
  const { connect, connectors } = useConnect();

  React.useEffect(() => {
    if (!isConnected) {
      connect({
        connector: connectors[0],
      });
    }
  }, [isConnected, connect, connectors]);

  if (!isConnected) {
    return null;
  }

  return <>{children}</>;
};

interface RenderHookOptions<Props> {
  initialProps?: Props;
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>;
}

export function renderHook<Result, Props>(
  renderCallback: (initialProps: Props) => Result,
  options: RenderHookOptions<Props> = {}
) {
  const { initialProps } = options;

  const result = React.createRef<Result>();

  function TestComponent({ renderCallbackProps }) {
    const pendingResult = renderCallback(renderCallbackProps);

    React.useEffect(() => {
      // @ts-ignore
      result.current = pendingResult;
    });

    return null;
  }

  const { rerender: baseRerender, unmount } = render(
    <TestComponent renderCallbackProps={initialProps} />
  );

  function rerender(rerenderCallbackProps) {
    return baseRerender(
      <TestComponent renderCallbackProps={rerenderCallbackProps} />
    );
  }

  return { result, rerender, unmount };
}

const { chains, provider, webSocketProvider } = configureChains(
  [chain.rinkeby],
  [
    // publicProvider(),
    jsonRpcProvider({
      rpc: (chain) => ({
        http: "http://127.0.0.1:8545",
      }),
    }),
  ]
);

const AllTheProviders: React.FC = ({ children }) => {
  function getSigner() {
    const p = new providers.JsonRpcProvider("http://127.0.0.1:8545");

    const wallet = new Wallet(
      "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
    );

    return p.getSigner(wallet.address);
  }

  const mockConnector = new MockConnector({
    chains,
    options: {
      chainId: 4,
      signer: getSigner(),
    },
  });

  const client = createClient({
    // autoConnect: true,
    connectors: [mockConnector],
    provider,
    webSocketProvider,
  });

  return <WagmiConfig client={client}>{children}</WagmiConfig>;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
