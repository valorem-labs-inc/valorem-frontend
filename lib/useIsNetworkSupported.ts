import { useEffect, useState } from "react";
import { useNetwork } from "wagmi";

export function useIsNetworkSupported() {
  const [isSupported, setIsSupported] = useState(false);

  const { activeChain } = useNetwork();

  useEffect(() => {
    if (!activeChain || activeChain.unsupported) {
      setIsSupported(false);
    } else {
      setIsSupported(true);
    }
  }, [activeChain]);

  return isSupported;
}
