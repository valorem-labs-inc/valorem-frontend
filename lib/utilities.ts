import { Contract, Signer } from "ethers";

import erc20ABI from "./abis/erc20";

export async function handleApproveToken(
  asset: string,
  spender: string,
  signer: Signer,
  callback?: (response: any) => void
): Promise<void> {
  const erc20Instance = new Contract(asset, erc20ABI, signer);

  if (!erc20Instance) {
    console.warn(`No erc20 instance for "${asset}" - no approval possible`);
    return;
  }

  const approvalTransaction = await erc20Instance.approve(
    spender,
    "115792089237316195423570985008687907853269984665640564039457584007913129639935"
  );

  return approvalTransaction.wait().then((approvalResponse) => {
    // this.setState({ needsApproval: false }, );
    if (callback) {
      return callback(approvalResponse);
    }
    return approvalResponse;
  });
}
