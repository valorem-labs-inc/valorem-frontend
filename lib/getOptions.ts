import { BigNumber, ethers, Signer } from "ethers";
import moment from "moment";

import graphql from "../graphql/client";
import { optionsByAccountQuery } from "../graphql/queries/options";
import getERC20Balance from "./getERC20Balance";
import type {
  GraphBalanceOption,
  GraphBalancesResponse,
  GraphTokenBalanceResponse,
  Option,
  OptionDetails,
} from "./types";
import unfreezeApolloCacheValue from "./unfreezeApolloCacheValue";

function graphOptionToOption(graphOption: GraphBalanceOption): Option {
  return {
    id: graphOption.id.toString(),
    exerciseAmount: BigNumber.from(graphOption.exerciseAmount),
    exerciseAsset: graphOption.exerciseAsset.id,
    exerciseTimestamp: parseInt(graphOption.exerciseTimestamp ?? "0", 10),
    expiryTimestamp: parseInt(graphOption.expiryTimestamp ?? "0", 10),
    underlyingAmount: BigNumber.from(graphOption.underlyingAmount),
    underlyingAsset: graphOption.underlyingAsset.id,
  };
}

async function getRawOptions(account: string): Promise<GraphBalancesResponse> {
  const q = {
    query: optionsByAccountQuery,
    variables: {
      account: account.toLowerCase(),
    },
  };

  const results = await graphql.query(q);

  if (!results) {
    console.error("no results", q);
    return { account: { ERC1155balances: [] } };
  }

  const { data } = results;
  if (!data) {
    console.log("empty options");
    return { account: { ERC1155balances: [] } };
  }

  console.log(`returning ${data.account.ERC1155balances.length} options`);
  return data as GraphBalancesResponse;
}

export default async function getOptions(account: string): Promise<Option[]> {
  const graphResponse = await getRawOptions(account);

  const balances: GraphTokenBalanceResponse[] =
    graphResponse.account.ERC1155balances;
  const options: Option[] = [];
  balances.forEach(({ token }: GraphTokenBalanceResponse) => {
    if (token.type === 1) {
      const option = unfreezeApolloCacheValue(token.option);
      options.push(graphOptionToOption(option as GraphBalanceOption));
    }
  });
  return options;
}

export async function getOptionsWithDetails(
  account: string,
  provider: Signer
): Promise<OptionDetails[]> {
  const graphResponse = await getRawOptions(account);

  let options: OptionDetails[] = [];
  const tokenBalances: Record<string, BigNumber> = {};
  const now = moment();

  const getAssetBalance = async (tokenId: string): Promise<BigNumber> => {
    if (!tokenBalances[tokenId]) {
      tokenBalances[tokenId] = await getERC20Balance(
        tokenId,
        account,
        provider
      );
    }
    return tokenBalances[tokenId];
  };

  if (graphResponse.account?.ERC1155balances) {
    // console.log(
    // "response",
    // JSON.stringify(graphResponse.account.ERC1155balances, null, 2)
    // );
    const type1Balances = graphResponse.account.ERC1155balances.filter(
      ({ token }) => token.type === 1
    );
    // console.log("type1Balances", type1Balances);
    const balanceQueries = type1Balances.map(({ token }) =>
      getAssetBalance(token.option.underlyingAsset.id)
    );

    const balances = await Promise.all(balanceQueries);

    options = type1Balances.map(({ token, valueExact }, index) => {
      const tokenBalance = balances[index];
      const graphOption = unfreezeApolloCacheValue(token.option);
      const option = graphOptionToOption(graphOption as GraphBalanceOption);
      const lastDate = moment.unix(option.expiryTimestamp);
      const firstDate = moment.unix(option.exerciseTimestamp);
      const canExercise =
        tokenBalance.gte(option.exerciseAmount) &&
        now.isBetween(firstDate, lastDate);
      return {
        balance: BigNumber.from(valueExact),
        option,
        canExercise,
        needsApproval: true,
      } as OptionDetails;
    });
    console.log(`returning ${options.length} options`);
    options.sort((a, b) => {
      return a.option.expiryTimestamp - b.option.expiryTimestamp;
    });
  }
  // return null;
  // console.log(`returning ${options.length} options`);
  return options;
}
