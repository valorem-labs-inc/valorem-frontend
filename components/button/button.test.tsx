import React from "react";
import { render, screen } from "../../lib/testing";
import Button from "./index";

describe("button component", () => {
  it("should render", () => {
    render(<Button theme="purple-blue">Hello</Button>);

    const btn = screen.getByText("Hello");

    expect(btn).toMatchSnapshot();
  });
});
