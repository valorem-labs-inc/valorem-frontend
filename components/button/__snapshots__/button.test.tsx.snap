// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button component should render 1`] = `
.c0 {
  border: none;
  padding: 10px 16px;
  border-radius: 0.7rem;
  font-size: 1rem;
  font-family: "Neue Montreal";
  --webkit-font-smoothing: antialiased;
  cursor: pointer;
  width: auto;
  background: var(--purple-blue);
  color: var(--white);
}

.c0:disabled {
  opacity: 0.6;
  pointer-events: none;
  background: var(--gray-500);
}

.c0:active {
  position: relative;
  top: 1px;
}

<button
  class="c0"
>
  Hello
</button>
`;
